homeassistant/components/minio/minio_helper.py

```diff
@@ -0,0 +1,208 @@
+"""Minio helper methods."""
+
+import collections
+import json
+import logging
+import re
+import threading
+from queue import Queue
+from typing import Iterator, List
+from urllib.parse import unquote
+
+from minio import Minio
+
+_LOGGER = logging.getLogger(__name__)
+
+_METADATA_RE = re.compile('x-amz-meta-(.*)', re.IGNORECASE)
+
+
+def normalize_metadata(metadata: dict) -> dict:
+    """Normalize object metadata by stripping the prefix."""
+    new_metadata = {}
+    for meta_key, meta_value in metadata.items():
+        match = _METADATA_RE.match(meta_key)
+        if not match:
+            continue
+
+        new_metadata[match.group(1).lower()] = meta_value
+
+    return new_metadata
+
+
+def get_minio_notification_response(
+        minio_client,
+        bucket_name: str,
+        prefix: str,
+        suffix: str,
+        events: List[str]
+):
+    """Start listening to minio events. Copied from minio-py."""
+    query = {
+        'prefix': prefix,
+        'suffix': suffix,
+        'events': events,
+    }
+    # noinspection PyProtectedMember
+    # pylint: disable=W0212
+    return minio_client._url_open(
```

Comment:
@MartinHjelmare 
Convo from previous PR.
https://github.com/minio/minio-py/blob/7107c84183cf5fb4deff68c0a16ab9f1c0b4c37e/minio/api.py#L491
Implementation in minio library wraps everything, and only yields retrieved events, and gives no access to the underlying request.
So there is no way to interrupt this, and cleanly stop the listener.

Reply:
It would be better to make a PR to the minio library with a change so it's possible to use the library cleanly.

Reply:
Indeed I agree. I will create PR there as well, and will update it here once it gets merged into upstream.